### PR TITLE
New App: SUPACLOCK Pro ULTRA

### PR DESCRIPTION
**Pull request by devsnd**
A new watch face with a bunch of clock-infos and a nice font

### CODE REVIEW

File: apps/supaclk/app.js [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+  var date = new Date();
+  let minutes = date.getMinutes();
+  minutes = (minutes < 10 ? '0' : '') + minutes;
+  let hours = date.getHours()+'';
+  g.drawImage(supaClockImg, pos.x, pos.y, {scale: 2});
+  g.setColor(0).setFont('6x8:2').setFontAlign(0, 1).drawString(hours + ':' + minutes, R.x2/2, pos.y + supaClockImg.height + 25)
+}
+
+// for fast startup-feeling, draw the splash screen once directly once
+drawSplashScreen(0, 20);
```
Comment (thyttan) [inline apps/supaclk/app.js:38]: I dig this!

File: apps/supaclk/app.js [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+let clockInfoDraw = (itm, info, options) => {
+  g.reset().setFontPlayfairDisplaySm().setColor(g.theme.fg).setBgColor(g.theme.bg);
+  if (options.focus) g.setBgColor("#FF0");
+  g.clearRect({x:options.x,y:options.y,w:options.w,h:options.h});
+
+  if (info.img) {
+    g.drawImage(info.img, options.x+1,options.y+2);
+  }
+  var text = info.text.toString().toUpperCase();
+  if (g.setFontPlayfairDisplaySm().stringWidth(text)+24-2>options.w) g.setFont("8x12");
```
Comment (thyttan) [inline apps/supaclk/app.js:121]: I get there errors, using firmware 2v22.5 (have not tried stable 2v22 for now):

```
Uncaught Error: Unknown font "8x12"
 at line 36 col 675 in supaclk.app.js
...>options.w)g.setFont("8x12");g.setFontAlign(-1,-1).drawStrin...
                              ^
in function "draw" called from line 11 col 3260 in clock_info
options.draw(itm,itm.get(),options);
                                  ^
in function "drawItem" called from line 11 col 3319 in clock_info
drawItem(itm)
            ^
in function called from system
 at line 36 col 675 in supaclk.app.js
...>options.w)g.setFont("8x12");g.setFontAlign(-1,-1).drawStrin...
                              ^
in function "draw" called from line 11 col 3260 in clock_info
options.draw(itm,itm.get(),options);
                                  ^
in function "drawItem" called from line 11 col 3319 in clock_info
drawItem(itm)
            ^
in function called from system
 at line 36 col 1155 in supaclk.app.js
...>options.w)g.setFont("8x12");g.setFontAlign(1,-1).drawString...
                              ^
in function "draw" called from line 11 col 3260 in clock_info
options.draw(itm,itm.get(),options);
                                  ^
in function "drawItem" called from line 11 col 3319 in clock_info
drawItem(itm)
            ^
in function called from system
```

Looking at [the `g.setFont` reference](https://www.espruino.com/ReferenceBANGLEJS2#l_Graphics_setFont) I don't think `"8x12"` is available. But you could do `g.setFont("Vector:12")` or `g.setFont("4x6:2")` to maybe achieve what you want?

File: apps/supaclk/app.js [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+let clockInfoDrawR = (itm, info, options) => {
+  g.reset().setFontPlayfairDisplaySm().setColor(g.theme.fg).setBgColor(g.theme.bg);
+  if (options.focus) g.setBgColor("#FF0");
+  g.clearRect({x:options.x,y:options.y,w:options.w,h:options.h});
+
+  if (info.img) {
+    g.drawImage(info.img, options.x + options.w-1-24,options.y+2);
+  }
+  var text = info.text.toString().toUpperCase();
+  if (g.setFontPlayfairDisplaySm().stringWidth(text)+24-2>options.w) g.setFont("8x12");
```
Comment (thyttan) [inline apps/supaclk/app.js:134]: ... also here `"8x12"` is not available to `setFont` as per review comment above.

File: apps/supaclk/app.js [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+// Show launcher when middle button pressed
+Bangle.setUI({
+  mode : "clock",
+  remove : function() {  // for fastloading, clear the app memory
+    if (drawTimeout) clearTimeout(drawTimeout);
+    drawTimeout = undefined;
+    delete Graphics.prototype.setFontPlayfairDisplay
+    delete Graphics.prototype.setFontPlayfairDisplaySm
+    delete clockInfoItems;
+    clockInfoMenu1.remove();
```
Comment (thyttan) [inline apps/supaclk/app.js:154]: If I press the HW button when the splash screen in showing the screen will be blank afterwards with this error message in the Web IDE, on fw 2v22.5:

```
 ____                 _
|  __|___ ___ ___ _ _|_|___ ___
|  __|_ -| . |  _| | | |   | . |
|____|___|  _|_| |___|_|_|_|___|
         |_| espruino.com
 2v22.5 (c) 2024 G.Williams
>
Uncaught Error: Can't read property 'remove' of undefined
 at line 155 col 19 in supaclk.app.js
    clockInfoMenu1.remove();
                  ^
in function "a" called from line 2 col 239
...e;delete Bangle.uiRemove;a()}g.reset();if(c){if("updown"==c)...
                              ^
in function called from line 25 col 338 in .boot0
...options.back;}setUI(mode,cb);
                              ^
at line 25 col 338 in .boot0
...options.back;}setUI(mode,cb);
                              ^
in function "setUI" called from line 1 col 31
Bangle.uiRemove?(Bangle.setUI(),(__FILE__=a)||(a=".bootcde")...
                              ^
in function called from line 15 col 1797 in .boot0
...);Bangle.load=orig;}else o(name);
                                  ^
at line 15 col 1797 in .boot0
...);Bangle.load=orig;}else o(name);
                                  ^
in function "load" called from line 1 col 408
...))}c?Bangle.load(b.launcher):eval('E.showMessage("No Launche...
                              ^
in function called from system
Uncaught Error: Unhandled promise rejection: Error: Function "setFontPlayfairDisplay" not found!
 at line 97 col 24 in supaclk.app.js
  g.setFontAlign(1, 0).setFontPlayfairDisplay().drawString(h...
                       ^
in function "draw" called from line 175 col 8 in supaclk.app.js
  draw();
       ^
in function called from system
> 
```
I tried this suggestion

```suggestion
    clockInfoMenu1&&clockInfoMenu1.remove();
```

... but it didn't play well. The splash screen would continue to draw over the loaded launcher and eventually clear the screen. The launcher still works though if I swipe the screen to interact. Maybe the splash screen could be made to stop short if the HW button is pressed? Maybe not for this PR if you don't want to. Anyway I subsequently also got this error message with the modified code:

```
 ____                 _
|  __|___ ___ ___ _ _|_|___ ___
|  __|_ -| . |  _| | | |   | . |
|____|___|  _|_| |___|_|_|_|___|
         |_| espruino.com
 2v22.5 (c) 2024 G.Williams
>
Uncaught Error: Unhandled promise rejection: Error: Function "setFontPlayfairDisplay" not found!
 at line 97 col 24 in supaclk.app.js
  g.setFontAlign(1, 0).setFontPlayfairDisplay().drawString(h...
                       ^
in function "draw" called from line 175 col 8 in supaclk.app.js
  draw();
       ^
in function called from system
Uncaught Error: Unhandled promise rejection: Error: Function "setFontPlayfairDisplay" not found!
 at line 97 col 24 in supaclk.app.js
  g.setFontAlign(1, 0).setFontPlayfairDisplay().drawString(h...
                       ^
in function "draw" called from line 175 col 8 in supaclk.app.js
  draw();
       ^
in function called from system
>
```

In conclusion, I don't have a short and neat solution to suggest. Maybe you could try catching the original error for now and do something if it occurs?

Comment (thyttan) [reply apps/supaclk/app.js:154]: Maybe do a check for if `setFontPlayfairDisplay` is defined and if it's not `return`?

Comment (thyttan) [reply apps/supaclk/app.js:154]: Ok, I took a dig at this and think I have solved it. I'll do a PR to your fork/branch if you want to merge it and have it go in with this PR to the official repo.

**Comment by thyttan** [comment]
After my tweaks, especially the HRM clock info doesn't behave perfectly, sometimes drawing outside of its rectangle depending on which font it uses.

**Comment by thyttan** [comment]
@devsnd did you also see the HRM clock info sometimes drawing outside its containing rectangle?

I could merge now but would prefer if we could get that sorted as well 🙂

**Comment by devsnd** [comment]
I am in no hurry to get it merged, but I also have very little time right now (my second son was born a few days ago and I do some hacking whenever I have a minute here or there).

In essence, I'd rather have it merged and do a v0.02 whenever I find some time, unless there are any bugs that break any normal user interaction. A smaller rendering bug is fine for me.

(I only today discovered at 20:00, that the main font is too large 🤦 , so that's something I'll also change in the next version)

**Comment by thyttan** [comment]
Congratulations to you and your family! 🎆

Yes - then lets merge :)

**Comment by thyttan** [comment]
Thanks!
